src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -246,38 +246,58 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
-          ReleaseTaskArgs(task_id);
-          // It may be that no worker was available with the correct runtime env or
-          // correct job ID.  However, another task with a different env or job ID
-          // might have a worker available, so continue iterating through the queue.
-          work_it++;
-          // Keep track of runtime env that doesn't have workers available so we
-          // won't call PopWorker for subsequent tasks requiring the same runtime env.
-          blocked_runtime_env_to_skip.insert(runtime_env_worker_key);
-          continue;
-        }
-
-        RAY_LOG(DEBUG) << "Dispatching task " << task_id << " to worker "
-                       << worker->WorkerId();
-        auto reply = std::get<1>(*work_it);
-        auto callback = std::get<2>(*work_it);
-        Dispatch(worker, leased_workers_, allocated_instances, task, reply, callback);
-      }
-
-      if (!spec.GetDependencies().empty()) {
-        task_dependency_manager_.RemoveTaskDependencies(
-            task.GetTaskSpecification().TaskId());
+        work->waiting_worker_popped = true;
+        pending_workers_index_.emplace(task_id, work);
+        worker_pool_.PopWorker(
+            spec,
+            [this, allocated_instances, task_id, runtime_env_worker_key](
+                const std::shared_ptr<WorkerInterface> worker, Status status) {
+              auto it = pending_workers_index_.find(task_id);
+              RAY_CHECK(it != pending_workers_index_.end());
+              const auto &work = it->second;
+              const auto &reply = work->reply;
+              const auto &callback = work->callback;
+              const auto &canceled = work->canceled;
+              if (canceled || !worker) {
+                if (canceled) {
+                  RAY_LOG(DEBUG)
+                      << "Task " << task_id
+                      << " has been canceled when worker popped, maybe the task has "
+                         "been cancelled.";
+                } else {
+                  RAY_LOG(DEBUG)
+                      << "This node has available resources, but no worker processes "
+                         "to grant the lease "
+                      << task_id << ", status = " << status.CodeAsString();
+                }
+                // We've already acquired resources so we need to release them to avoid
+                // double-acquiring when the next invocation of this function tries to
+                // schedule this task.
+                cluster_resource_scheduler_->ReleaseWorkerResources(allocated_instances);
+                // No worker available, we won't be able to schedule any kind of task.
+                // Worker processes spin up pretty quickly, so it's not worth trying to
+                // spill this task.
+                ReleaseTaskArgs(task_id);
+                // TODO(guyang.sgy): re-schedule the task by status code.
+                work->waiting_worker_popped = false;
```

Comment:
Couldn't this hang? Since there is no guarantee DispatchScheduledTasksToWorkers will get called again.

Reply:
Good comment! If the status code is JobNotStarted or WorkerPendingRegistration,  maybe `DispatchScheduledTasksToWorkers ` will be triggered by other events? Such as job register or worker register. But If the status code is RuntimeEnvCreationFailed, this will hang. I should do some work here. What do you think?

Reply:
Hmm now that I'm thinking about this more, this might be okay since the previous version of the code also didn't handle the case.

Could you add a `NOTE:` comment to explain that it's assumed that DispatchScheduledTasksToWorkers will get called again?

Reply:
I have solved this problem now. If `RuntimeEnvCreationFailed `happens, we will fail the task directly.